tsdb/db.go

```diff
@@ -995,11 +1003,24 @@ func (db *DB) run(ctx context.Context) {
 
 			db.autoCompactMtx.Lock()
 			if db.autoCompact {
-				if err := db.Compact(ctx); err != nil {
-					level.Error(db.logger).Log("msg", "compaction failed", "err", err)
-					backoff = exponential(backoff, 1*time.Second, 1*time.Minute)
-				} else {
-					backoff = 0
+				// TODO: refactor?
+				if db.head.compactable() {
+					if compactShiftStart.IsZero() {
+						// Start counting for the offset.
+						compactShiftStart = time.Now()
+					}
+					if time.Since(compactShiftStart) > compactOffset {
+						if err := db.Compact(ctx); err != nil {
+							level.Error(db.logger).Log("msg", "compaction failed", "err", err)
+							backoff = exponential(backoff, 1*time.Second, 1*time.Minute)
+						} else {
+							backoff = 0
+							compactShiftStart = time.Time{}
+						}
+					}
+				} else if !compactShiftStart.IsZero() {
+					// In case the compaction was run manually
```

Comment:
I was a bit confused by this.  I think you mean "we were waiting to compact, and yet the head is not compactable, so assume it has been compacted by another route in the meantime"; is that right?